gh-push

```diff
@@ -1,36 +1,91 @@
 #!/usr/bin/env bash
 
-HEMLOCK_PRE_PUSH_CLONE_PATH=.pre_push_clone
-_CWD=$(pwd)
+set -e
 
-trap "rm -rf ${HEMLOCK_PRE_PUSH_CLONE_PATH} && cd ${_CWD}" EXIT
-trap "exit 1" SIGINT
-pushd $(git rev-parse --show-toplevel) > /dev/null || (echo "Not in a git repo."; exit 1)
+_DRY_RUN=0
+for _ARG in "$@"
+do
+    if [ "${_ARG}" = "--dry-run" ] || [ "${_ARG}" = "-n" ]
+    then
+        echo "Dry run."
+        _DRY_RUN=1
+        break
+    fi
+done
 
-git push $@
-if [ $? -ne 0 ]
+_TREE=$(git cat-file -p HEAD | head -n 1)
+_SHA=$(git rev-parse HEAD)
+_INTERRUPTED=0
+trap "_INTERRUPTED=1" SIGINT
+
+# Arg 1 is GitHub API commit status.
+# Arg 2 is additional description text.
+set_status() {
+    if [ ${_DRY_RUN} -ne 0 ]
+    then
+        return
+    fi
+    
+    _DESCRIPTION="$2, ${_TREE}"
+    _GH_API_RET=0
+    while :
+    do
+        if [ ${_INTERRUPTED} -ne 0 ]
+        then
+            exit 1
+        elif [ ${_GH_API_RET} -ne 0 ]
+        then
+            _GH_API_RET=0
+            echo "Waiting 3 seconds to retry set GitHub commit status to \"$1\"."
+            sleep 3
+        else
+            gh api /repos/{owner}/{repo}/statuses/${_SHA} -f state="$1" -f context="Hemlock Push" \
+              -f description="${_DESCRIPTION}" > /dev/null || _GH_API_RET=1
```

Comment:
We can't just keep pushing statuses. It's fairly unlikely in a normal workflow, but the list of statuses is saved and never truncated. Eventually, we hit a limit.

Reply:
I wasn't worried about the limit because docs suggest it's 1000 per commit/context pair. The docs also don't show a way to update the status.

We hit a limit at 30 statuses for a commit.

Reply:
Not true. We actually did hit the limit. While developing this script, I accidentally left a three zombie processes around that were committing a fail status every 3 seconds all night. I'd intentionally created those looping processes, but I didn't realize I'd left them running.